### PR TITLE
Implement configuration validation

### PR DESCRIPTION
**Pull request by lorenzoc25**
Implement a middleware to check if the configuration object is valid. If any of the required fields is missing, raise an error.

### CODE REVIEW

File: src/validate.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+) => {
+  const { options } = context;
+  const {
+    upstream,
+    firewall,
+    error,
+    // cors,
+    // optimization,
+    // header,
+    // security,
```
Comment (xiaoyang-sde) [inline src/validate.ts:17]: I think these comments could be removed

File: src/validate.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  } = options;
+
+  if (upstream === undefined) {
+    throw new Error('Invalid upstream configuration. Please check again that you did not input empty value.');
```
Comment (xiaoyang-sde) [inline src/validate.ts:21]: May change to `The required field 'upstream' in the configuration is missing`.

File: src/validate.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  if (Array.isArray(upstream)) {
+    upstream.forEach((userConfig) => {
+      if (userConfig.domain === undefined) {
+        throw new Error('Invalid domain configuration. Please check that you included domain in the configuration');
```
Comment (xiaoyang-sde) [inline src/validate.ts:27]: The required field 'domain' in the `upstream` configuration is missing.

File: src/validate.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    // header,
+    // security,
+  } = options;
+
+  if (upstream === undefined) {
+    throw new Error('Invalid upstream configuration. Please check again that you did not input empty value.');
+  }
+
+  if (Array.isArray(upstream)) {
+    upstream.forEach((userConfig) => {
```
Comment (xiaoyang-sde) [inline src/validate.ts:25]: `userConfig` -> `upstreamOptions`

File: src/validate.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      if (userConfig.domain === undefined) {
+        throw new Error('Invalid domain configuration. Please check that you included domain in the configuration');
+      }
+    });
+  } else if (upstream.domain === undefined) {
+    throw new Error('Invalid domain configuration. Please check that you included domain in the configuration');
+  }
+
+  const checkFirewall = (userConfig: FirewallOptions) : void => {
```
Comment (xiaoyang-sde) [inline src/validate.ts:34]: This function could be moved outside of `useValidate`. The parameters should be placed in a new line. Please use an informative name for the function such as `validateFirewallOptions`.
```
const validateFirewallOptions = (
  firewallOptions: FirewallOptions,
): void =>
```

File: src/validate.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  if (firewall !== undefined) {
+    if (Array.isArray(firewall)) {
+      firewall.forEach(checkFirewall);
+    } else {
+      checkFirewall(firewall);
+    }
+  }
+
+  const checkError = (userConfig : ErrorOptions) : void => {
```
Comment (xiaoyang-sde) [inline src/validate.ts:64]: This function could be moved outside of `useValidate`.

File: src/validate.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    if (userConfig.errorCode === undefined || userConfig.responsePath === undefined) {
+      throw new Error('Invalid customized error configuration. ');
```
Comment (xiaoyang-sde) [inline src/validate.ts:66]:
```suggestion
      throw new Error('Invalid customized error configuration.');
```

**Review by xiaoyang-sde — CHANGES_REQUESTED** [review]

**Comment by xiaoyang-sde** [comment]
The unit test could be removed if it's not easy to implement.

**Comment by lorenzoc25** [comment]
Thanks for the suggesting! I've removed the test cases and re-format the file in the newest commit :)

**Review by xiaoyang-sde — APPROVED** [review]

**Pull request merged**